service/worker/watchdog/workflow.go

```diff
@@ -0,0 +1,169 @@
+		MaxCount:        1000,
+		TTL:             24 * time.Hour,
+		Pin:             false,
+	}
+
+	w := Workflow{
+		watchdog:  wd,
+		processed: cache.New(cacheOpts),
+	}
+
+	workflow.RegisterWithOptions(w.workflowFunc, workflow.RegisterOptions{Name: watchdogWFTypeName})
+	activity.RegisterWithOptions(w.handleCorrputedWorkflow, activity.RegisterOptions{Name: handleCorrputedWorkflowActivity})
+}
+
+// workflowFunc is the workflow that performs actions for WatchDog
+func (w *Workflow) workflowFunc(ctx workflow.Context) error {
+	requestCh := workflow.GetSignalChannel(ctx, CorruptWorkflowWatchdogChannelName)
+	logger := w.watchdog.logger
+
+	for {
+		var request CorruptWFRequest
+		if more := requestCh.Receive(ctx, &request); !more {
+			logger.Info("Corrupt workflow channel closed")
+			return cadence.NewCustomError("signal_channel_closed")
+		}
+
+		if w.watchdog.config.CorruptWorkflowWatchdogPause() {
+			logger.Warn("Corrupt workflow execution is paused. Enable to continue processing")
+			continue
+		}
+		opt := workflow.WithActivityOptions(ctx, handleCorruptWorkflowOptions)
+		_ = workflow.ExecuteActivity(opt, handleCorrputedWorkflowActivity, request).Get(ctx, nil)
+	}
+}
+
+func (w *Workflow) getProcessedID(request *CorruptWFRequest) string {
+	return fmt.Sprintf("%s-%s", request.Workflow.WorkflowID, request.Workflow.RunID)
+}
+
+// handleCorrputedWorkflowActivity is activity to handle corrupted workflows in DB
+func (w *Workflow) handleCorrputedWorkflow(ctx context.Context, request *CorruptWFRequest) error {
+	if w.processed.Get(w.getProcessedID(request)) != nil {
+		// We already processed this workflow before and couldn't decide if we should delete
+		return nil
+	}
+
+	logger := activity.GetLogger(ctx).With(
+		zap.String("DomainName", request.DomainName),
+		zap.String("WorkflowID", request.Workflow.GetWorkflowID()),
+		zap.String("RunID", request.Workflow.GetRunID()))
+	logger.Info("Watchdog processing possible corrupt workflow")
+	domainEntry, err := w.watchdog.domainCache.GetDomain(request.DomainName)
+	if err != nil {
+		logger.Error("Failed to get domain entry", zap.Error(err))
+		return err
+	}
+
+	tagged := w.watchdog.scopedMetricClient.Tagged(
+		metrics.DomainTag(domainEntry.GetInfo().Name),
+	)
+	clusterName := domainEntry.GetReplicationConfig().ActiveClusterName
+	adminClient := w.watchdog.clientBean.GetRemoteAdminClient(clusterName)
+	maintainWFRequest := &types.AdminMaintainWorkflowRequest{
+		Domain:     request.DomainName,
+		Execution:  &request.Workflow,
+		SkipErrors: true,
+	}
+	_, err = adminClient.MaintainCorruptWorkflow(ctx, maintainWFRequest)
+	if err != nil {
+		return err
+	}
+
+	tagged.AddCounter(metrics.WatchDogNumCorruptWorkflowProcessed, 1)
+
+	// We couldn't decide if we should delete this workflow. That means the same logs will continue
+	// and the workflow will keep getting signals. So adding workflow to a cache to skip processing
+	w.processed.Put(w.getProcessedID(request), struct{}{})
```

Comment:
I didn't quite get this part. Could you elaborate a bit more?

Reply:
This was to avoid spamming with retries. I will move this to the client (mentioned above).